doc/designs/fips_indicator.md

```diff
@@ -26,59 +26,122 @@ The following information was extracted from the FIPS 140-3 IG [5] “2.4.C Appr
 
 The following rules will apply to any code that currently is not FIPS approved,
 but needs to be.
 
-- There will be a 'fips approved' mode per provider operation context that is on by default.
-- In this mode any FIPS rule that is violated will result in an error.
-- The 'fips approved' mode may be turned off per context via a OSSL_PARAM setter, if it
-is turned off then the operation is not FIPS approved.
-- A getter is required that returns if the operation is 'fips approved' using an
-OSSL_PARAM. The returned value can be a combination of the set 'fips_approved' mode
-plus any other logic. This is an explicit indicator.
-- Any algorithm that is transitioning [4] to not being allowed should be removed from
-the fips provider.
-- If an algorithm is transitioning [4] to be only allowed for processing
-(e.g. verification, signing, key validation) then the protection code
-(keygen, signing, encryption) should be removed from the fips provider, any
-attempt to use the protection API's should result in an error. The processing
-code should still be functional.
-
-Other rules:
-The existing flags that we set in the fips config file to control security checks
-will continue to function as they do now, and will not be affected by the
-strict mode variable.
-The getter to determine if we are fips approved will however take the flags
-into account.
-
-Motivation
+- The fipsinstall application will have a configurable item added for each algorithm that requires a change.
+These options will be passed to the FIPS provider in a manner similar to existing code.
+
+- A user defined callback similar to OSSL_SELF_TEST will be added. This callback
+will be triggered whenever an approved mode test fails.
+It may be set up by the user using
+
+```c
+OSSL_INDICATOR_set_callback(OSSL_LIB_CTX *libctx, OSSL_CALLBACK *cb, void *cbarg)
+```
+
+- The callback API will be
+
+```c
+int OSSL_INDICATOR_callback(OSSL_LIB_CTX *libctx, const char *algtype,
+                            const char *algdesc)
+```
+
+This can be used during application testing to log that an indicator was
+triggered. The user callback may return 0 if the application wants an error
+to occur based on the indicator type and description.
+
+- To control an algorithm context's checks via code requires a setter (e.g OSSL_ALG_PARAM_STRICT_CHECKS),
+
+```c
+    p = OSSL_PARAM_locate_const(params, OSSL_ALG_PARAM_STRICT_CHECKS);
+    if (p != NULL
+        && !OSSL_PARAM_get_int(p, &ctx->strict_checks))
+        return 0;
+```
+
+The setter is initially -1 (unknown) and can be set to 0 or 1 via a set_ctx call.
+If the setter is -1, when the FIPS related approved mode check is done then it
+uses the value from the FIPS configuration instead.
+
+- To access the indicator via code requires a getter (e.g OSSL_ALG_PARAM_APPROVED_INDICATOR),
+
+```c
+    p = OSSL_PARAM_locate(params, OSSL_ALG_PARAM_APPROVED_INDICATOR);
+    if (p != NULL && !OSSL_PARAM_set_int(p, ctx->approved))
```

Comment:
Same as above, text indicates a getter is required, but the sample code illustrates a setter.

Reply:
This one sets the value into p and returns it.. (Thereby getting the value)